### PR TITLE
Update API Spec

### PR DESCRIPTION
**Pull request by decodingahmed**

### CODE REVIEW

File: src/MockClassifier.Api/Models/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿namespace MockClassifier.Api.Models
+{
+    public static class Constants
```
Comment (guypritchard) [inline src/MockClassifier.Api/Models/Constants.cs:3]: A lot of this is going to be repeated - it's probably time for us to think about a common library for this kind of thing.

File: src/MockClassifier.UnitTests/Extensions/MockHttpClientFactoryExtensions.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ internal static class MockHttpClientFactoryExtensions
                 .Returns(() =>
                 {
                     var client = messageHandler.ToHttpClient();
-                    client.BaseAddress = new Uri("https://dmr.fakeurl.com");
+                    client.BaseAddress = new Uri("https://tablelogapi.azurewebsites.net/app/fromclassifier");
```
Comment (guypritchard) [inline src/MockClassifier.UnitTests/Extensions/MockHttpClientFactoryExtensions.cs:19]: Is this URL a real service you're using?

File: src/MockClassifier.Api/Controllers/MessagesController.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            _naturalLanguageService = naturalLanguageService;
+            _encodingService = encodingService;
+        }
+
+        /// <summary>
+        /// Processes a string to identify classifications and issues call backs to Dmr for each classification.
+        /// </summary>
+        /// <param name="input">Property which contains the base64 encoded payload containing the message to be classified</param>
+        /// <returns>An empty 202/Accepted result</returns>
+        [HttpPost]
```
Comment (guypritchard) [inline src/MockClassifier.Api/Controllers/MessagesController.cs:32]:
```suggestion
        [HttpPost]
        [ProducesResponseType(StatusCodes.Status202Accepted)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
```

@yahya130 suggested this for the Dmr review.

**Review by guypritchard — APPROVED** [review]
Nice work!  We probably should revisit the shared library discussion sooner rather than later.

File: src/MockClassifier.Api/Controllers/MessagesController.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            if (string.IsNullOrEmpty(input))
+            {
+                return BadRequest(ModelState);
+            }
+            var decodedInput = _encodingService.DecodeBase64(input);
+            var payload = JsonSerializer.Deserialize<DmrRequestPayload>(decodedInput);
+
+            // Classify
+            List<string> classifications = _naturalLanguageService.Classify(payload.Message).ToList();
+            classifications = classifications.Concat(_tokenService.Classify(payload.Message).ToList()).ToList();
```
Comment (yahya130) [inline src/MockClassifier.Api/Controllers/MessagesController.cs:47]:
```suggestion
classifications.AddRange(_tokenService.Classify(payload.Message).ToList());
```

File: src/MockClassifier.Api/Controllers/MessagesController.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        /// <summary>
+        /// Builds a DmrRequest object from parameters
+        /// </summary>
+        /// <param name="message">The message to go into the .Payload.Message property</param>
+        /// <param name="classification">The classification to go into the .Payload.Classification property</param>
+        /// <param name="headers">The header collection from the origional Request. Used to create the .Header object</param>
+        /// <returns>A DmrRequest object</returns>
+        private static DmrRequest GetDmrRequest(string message, string classification, IHeaderDictionary headers)
```
Comment (yahya130) [inline src/MockClassifier.Api/Controllers/MessagesController.cs:66]: Thoughts on moving this logic into another class outside of the controller? would be nice for controller to only display HTTP/endpoint/route related methods only

File: src/MockClassifier.Api/Services/Dmr/DmrService.cs [hunk 6]
```diff
@@ -40,17 +50,31 @@ public async Task ProcessRequestsAsync()
...
                 {
-                    var response = await httpClient.PostAsJsonAsync("/", request).ConfigureAwait(true);
-                    _ = response.EnsureSuccessStatusCode();
+                    // Setup content
+                    var jsonPayload = JsonSerializer.Serialize(request.Payload);
+                    var jsonPayloadBase64 = _encodingService.EncodeBase64(jsonPayload);
+                    using var content = new StringContent(jsonPayloadBase64, Encoding.UTF8, MediaTypeNames.Application.Json);
 
+                    // Setup message
+                    using var requestMessage = new HttpRequestMessage()
```
Comment (yahya130) [inline src/MockClassifier.Api/Services/Dmr/DmrService.cs:59]: Might be worth moving this requestMessage creation logic into a separate method

Comment (decodingahmed) [reply src/MockClassifier.Api/Services/Dmr/DmrService.cs:59]: Yup, good spot. I've moved it into another method.

File: src/MockClassifier.UnitTests/Extensions/MockHttpClientFactoryExtensions.cs (hunk 2 above)
Comment (decodingahmed) [reply src/MockClassifier.UnitTests/Extensions/MockHttpClientFactoryExtensions.cs:19]: Yes, it is, this is a service that Martin is hosting that keeps an inmemory log of calls being made to it and in turns makes callbacks also.  Good way of debugging behaviour.

File: src/MockClassifier.Api/Controllers/MessagesController.cs (hunk 5 above)
Comment (decodingahmed) [reply src/MockClassifier.Api/Controllers/MessagesController.cs:66]: Usually I would agree with you but this is just a mock service so if it's okay with you I'd like to leave this as is. Is that all right?

File: src/MockClassifier.Api/Models/Constants.cs (hunk 1 above)
Comment (decodingahmed) [reply src/MockClassifier.Api/Models/Constants.cs:3]: 💯

File: src/MockClassifier.Api/Controllers/MessagesController.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            // Setup headers
+            _ = headers.TryGetValue(Constants.SentByHeaderKey, out var sentByHeader);
+            _ = headers.TryGetValue(Constants.MessageIdHeaderKey, out var messageIdHeader);
+            _ = headers.TryGetValue(Constants.SendToHeaderKey, out var sendToHeader);
+            _ = headers.TryGetValue(Constants.MessageIdRefHeaderKey, out var messageIdRefHeader);
+            var dmrHeaders = new Dictionary<string, string>
+            {
+                { Constants.SentByHeaderKey, sendToHeader },
+                { Constants.MessageIdHeaderKey, messageIdHeader },
```
Comment (tonistuulNG) [inline src/MockClassifier.Api/Controllers/MessagesController.cs:76]: am I reading this correctly that currently the mapping is: sentBy -> sendTo, sendTo -> sentBy, messageId -> messageId, messageIdRef -> messageIdRef?
If so, then response messageId should be new Guid() and messageIdRef should be request.messageId.

Comment (tonistuulNG) [reply src/MockClassifier.Api/Controllers/MessagesController.cs:76]: Also, if you could just reflect back X-Model-Type as well, that would be great - havent had time to update it in specs yet :(

File: src/MockClassifier.Api/Models/Constants.cs (hunk 1 above)
Comment (martinkearn) [reply src/MockClassifier.Api/Models/Constants.cs:3]: I've added this issue to track thus. P3 at the moment. https://github.com/buerokratt/Cross-functional-requirements/issues/46

File: src/MockClassifier.Api/Services/Dmr/DmrService.cs [hunk 8]
```diff
@@ -40,19 +44,41 @@ public async Task ProcessRequestsAsync()
...
 
-                    // Not sure how to resolve rule CA1848 so removing logging for now
-                    //logger.LogInformation($"Callback to DMR. Ministry = {request.Payload.Ministry}, Messages = {string.Join(", ", request.Payload.Messages)}");
+                    _logger.DmrCallback(request.Payload.Classification, request.Payload.Message);
                 }
-                catch (HttpRequestException)
+                catch (HttpRequestException exception)
                 {
-                    // Not sure how to resolve rule CA1848 so removing logging for now
-                    //logger.LogError(exception, "Call to DMR Service failed");
```
Comment (martinkearn) [inline src/MockClassifier.Api/Services/Dmr/DmrService.cs:53]: Now that the logger has been correctly setup, can we remove these comments and commented code? Applies throughout

File: src/MockClassifier.UnitTests/Extensions/MockHttpClientFactoryExtensions.cs (hunk 2 above)
Comment (martinkearn) [reply src/MockClassifier.UnitTests/Extensions/MockHttpClientFactoryExtensions.cs:19]: Yes, it is this which we can use until we have a better way of int testing: https://github.com/martinkearn/Table-Log-Api

**Review by martinkearn — APPROVED** [review]
Thanks for picking this up and getting it over the line @decodingahmed

**Review by yahya130 — APPROVED** [review]

File: src/MockClassifier.Api/Controllers/MessagesController.cs (hunk 5 above)
Comment (decodingahmed) [reply src/MockClassifier.Api/Controllers/MessagesController.cs:66]: Yahya and I discussed offline and we're happy to leave it as is. Thanks!

File: src/MockClassifier.Api/Controllers/MessagesController.cs (hunk 7 above)
Comment (decodingahmed) [reply src/MockClassifier.Api/Controllers/MessagesController.cs:79]: I've added the `X-Model-Type` functionality in

**Pull request merged**